[SPARK-7389] [core]Tachyon integration improvement

shimingfei opened the pull request:
Two main changes:

Add two functions in ExternalBlockManager, which are putValues and getValues
because the implementation may not rely on the putBytes and getBytes

improve Tachyon integration.
Currently, when putting data into Tachyon, Spark first serialize all data in one partition into a ByteBuffer, and then write into Tachyon, this will uses much memory and increase GC overhead

when get data from Tachyon, getValues depends on getBytes, which also read all data into On heap byte arry, and result in much memory usage.
This PR changes the approach of the two functions, make them read / write data by stream to reduce memory usage.

In our testing,  when data size is huge, this patch reduces about 30% GC time and 70% full GC time, and total execution time reduces about 10%

Commit:
add getValues and putValues in ExternalBlockManager's Interface

Commit:
Make getValues read data from Tachyon by stream
Make putValues write data to Tachyon by stream

AmplabJenkins commented:
Merged build triggered.

AmplabJenkins commented:
Merged build started.

SparkQA commented:
[Test build #31866 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/31866/consoleFull) for   PR 5908 at commit [`a6ce31e`](https://github.com/apache/spark/commit/a6ce31e2456cefe3ea362b76fcf7ff4bfad003ea).

srowen commented:
@shimingfei please review https://cwiki.apache.org/confluence/display/SPARK/Contributing+to+Spark  You will need a JIRA.

CC @calvinjia

SparkQA commented:
[Test build #31866 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/31866/consoleFull) for   PR 5908 at commit [`a6ce31e`](https://github.com/apache/spark/commit/a6ce31e2456cefe3ea362b76fcf7ff4bfad003ea).
 * This patch **passes all tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/31866/
Test PASSed.

shimingfei commented:
@haoyuan @zhzhan

zhzhan commented on core/src/main/scala/org/apache/spark/storage/ExternalBlockStore.scala:
```diff
@@ -62,42 +62,50 @@ private[spark] class ExternalBlockStore(blockManager: BlockManager, executorId:
-    logDebug(s"Attempting to write values for block $blockId")
-    val bytes = blockManager.dataSerialize(blockId, values)
-    putIntoExternalBlockStore(blockId, bytes, returnValues)
+    putIntoExternalBlockStore(blockId, Left(values), returnValues)
```
Need to take care the case that plugin does not implement putValue

zhzhan commented on core/src/main/scala/org/apache/spark/storage/ExternalBlockStore.scala:
```diff
@@ -46,7 +46,7 @@ private[spark] class ExternalBlockStore(blockManager: BlockManager, executorId:
 
   override def putBytes(blockId: BlockId, bytes: ByteBuffer, level: StorageLevel): PutResult = {
-    putIntoExternalBlockStore(blockId, bytes, returnValues = true)
+    putIntoExternalBlockStore(blockId, Right(bytes), returnValues = true)
```
Need to handle the case that plugin does not handle putBytes

zhzhan commented on core/src/main/scala/org/apache/spark/storage/ExternalBlockStore.scala:
```diff
@@ -113,7 +121,13 @@ private[spark] class ExternalBlockStore(blockManager: BlockManager, executorId:
   override def getValues(blockId: BlockId): Option[Iterator[Any]] = {
-    getBytes(blockId).map(buffer => blockManager.dataDeserialize(blockId, buffer))
+    try {
+      externalBlockManager.flatMap(_.getValues(blockId))
```
Need to handle the case plugin does not have getValues (getBytes only)

rxin commented on core/src/main/scala/org/apache/spark/storage/ExternalBlockStore.scala:
```diff
@@ -62,42 +62,50 @@ private[spark] class ExternalBlockStore(blockManager: BlockManager, executorId:
   private def putIntoExternalBlockStore(
       blockId: BlockId,
-      bytes: ByteBuffer,
+      data: Either[Iterator[_], ByteBuffer],
```
Please don't use Either. I think it's easier to just break this into two functions.

zhzhan commented:
@shimingfei  Some high level comments. I think putValue and putBytes, getValues and getBytes are overlapping. putValue/getValue is an improvement. Certainly plugin can implement both of them, but the shim layer does not need to enforce it. Probably a fallback mechanism should be provided. For example, putValues can have default implementation using putBytes, getValues can have default implementation getBytes. putBytes/getBytes can follow the similar way. Thus, user can implement all functions, or choose to fallback to the default.

Commit:
modify according to comments
split putIntoExternalBlockStore into two functions
add default implementation for getValues and putValues

AmplabJenkins commented:
Merged build triggered.

AmplabJenkins commented:
Merged build started.

SparkQA commented:
[Test build #31952 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/31952/consoleFull) for   PR 5908 at commit [`219e819`](https://github.com/apache/spark/commit/219e8197277c288be83edfb64350dccdb01a1441).

shimingfei commented:
@rxin  @zhzhan 
I have modified the code according to your comments, Thanks!
I think putBytes and getBytes are basic functions which external blockStores should offer. They are used to write/get serialized data to/from external blockStore

SparkQA commented:
[Test build #31952 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/31952/consoleFull) for   PR 5908 at commit [`219e819`](https://github.com/apache/spark/commit/219e8197277c288be83edfb64350dccdb01a1441).
 * This patch **passes all tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/31952/
Test PASSed.

zhzhan commented:
LGTM

calvinjia commented:
+1 LGTM, this should significantly help performance.

shimingfei commented:
@rxin Do you have more comments on this?

haoyuan commented:
+1 LGTM.

srowen commented on core/src/main/scala/org/apache/spark/storage/TachyonBlockManager.scala:
```diff
@@ -95,8 +94,23 @@ private[spark] class TachyonBlockManager() extends ExternalBlockManager with Log
-    os.close()
+    try {
+      os.write(bytes.array())
+      os.close()
```
Should the `close()` happen in a `finally` block? here and below. This also doesn't cancel in case of a runtime exception, and also doesn't propagate the exception or even log it.

shimingfei replied:
close() does not have to be in a finally block, because cancel() will also close the block file.
This will cancel the block when a runtime exception happens.
I will add log for it. Thanks!

srowen replied:
There is a code path where neither is cancelled though -- any subclass of `RuntimeException` or `Error` will invoke neither `cancel()` nor `close()`. If `close()` is safely idempotent then it should be in a `finally` block to catch all cases.

shimingfei replied:
The logic here is that: if any exception happens in write() or close(),  cancel the block file. And current code is able to cache all exceptions.
The OutStream is realized in Tachyon, not common OutStream in Java.

srowen replied:
I understand that. That is not my point. Any Error for example is not caught by this. This is why a close happens on the finally path by convention.

srowen commented on core/src/main/scala/org/apache/spark/storage/TachyonBlockManager.scala:
```diff
@@ -120,6 +134,16 @@ private[spark] class TachyonBlockManager() extends ExternalBlockManager with Log
+      return None
+    }
+    val is = file.getInStream(ReadType.CACHE)
+    assert (is != null)
```
I think the code is pretty inconsistent about this, but do you mean `assert` here? if assertions aren't on is it OK if this isn't checked? probably, just asking.  Also tiny nit: you don't need `return` below.

shimingfei replied:
I will remove assert and return.

srowen commented on core/src/main/scala/org/apache/spark/storage/ExternalBlockStore.scala:
```diff
@@ -113,7 +143,13 @@ private[spark] class ExternalBlockStore(blockManager: BlockManager, executorId:
+      externalBlockManager.flatMap(_.getValues(blockId))
+    } catch {
+      case NonFatal(t) =>
+        logError(s"error in getValues from $blockId", t)
```
This is really opinion too, but if it's not fatal, I think it could be a warning instead of error.

shimingfei replied:
use logError to keep consistent with other part of the code in this file

Commit:
modify according to comments, some code style change

AmplabJenkins commented:
Merged build triggered.

AmplabJenkins commented:
Merged build started.

SparkQA commented:
[Test build #32451 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/32451/consoleFull) for   PR 5908 at commit [`ac2aed3`](https://github.com/apache/spark/commit/ac2aed36afea19476f78613ee7fff8825d32159a).

SparkQA commented:
[Test build #32451 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/32451/consoleFull) for   PR 5908 at commit [`ac2aed3`](https://github.com/apache/spark/commit/ac2aed36afea19476f78613ee7fff8825d32159a).
 * This patch **passes all tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/32451/
Test PASSed.

rxin commented:
@srowen can you review this and ping me when you feel it is ready? I will take a final look after that.

srowen commented:
@rxin I'm fine with this excepting the potential point about closing resources in case of an `Error` at https://github.com/apache/spark/pull/5908#discussion_r30031320

Commit:
put close() in a finally block

shimingfei commented:
@srowen 
I have modified the code according to your comments.

AmplabJenkins commented:
Merged build triggered.

AmplabJenkins commented:
Merged build started.

SparkQA commented:
[Test build #32563 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/32563/consoleFull) for   PR 5908 at commit [`5361406`](https://github.com/apache/spark/commit/53614061ad7475b8618a8f666277004b5de02d97).

SparkQA commented:
[Test build #32563 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/32563/consoleFull) for   PR 5908 at commit [`5361406`](https://github.com/apache/spark/commit/53614061ad7475b8618a8f666277004b5de02d97).
 * This patch **passes all tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/32563/
Test PASSed.

srowen commented:
@rxin over to you. This looks good. I dont' know quite enough about the API and Tachyon to evaluate the semantics, but the code changes look good on the surface.

shimingfei commented:
@rxin  Request for your final review.

rxin commented on core/src/main/scala/org/apache/spark/storage/ExternalBlockManager.scala:
```diff
@@ -32,6 +32,8 @@ import java.nio.ByteBuffer
  */
 private[spark] abstract class ExternalBlockManager {
 
+  var blockManager: BlockManager = _
```
private?

shimingfei replied:
make it protected, because its sub-class may also use this variable.

rxin commented on core/src/main/scala/org/apache/spark/storage/ExternalBlockStore.scala:
```diff
@@ -62,42 +62,72 @@ private[spark] class ExternalBlockStore(blockManager: BlockManager, executorId:
+          blockId, Utils.bytesToString(size), finishTime - startTime))
+        PutResult(size, data)
+      } else {
+        logError(s"error in putBytes $blockId")
```
what's the error?

rxin commented on core/src/main/scala/org/apache/spark/storage/ExternalBlockStore.scala:
```diff
@@ -62,42 +62,72 @@ private[spark] class ExternalBlockStore(blockManager: BlockManager, executorId:
+      blockId: BlockId,
+      bytes: ByteBuffer,
+      returnValues: Boolean): PutResult = {
+    logDebug(s"Attempting to put block $blockId into ExtBlk store")
```
trace level

rxin replied:
ExtBlk -> external block

rxin commented on core/src/main/scala/org/apache/spark/storage/ExternalBlockStore.scala:
```diff
@@ -62,42 +62,72 @@ private[spark] class ExternalBlockStore(blockManager: BlockManager, executorId:
+          blockId, Utils.bytesToString(size), finishTime - startTime))
+        PutResult(size, data)
       } else {
         logError(s"error in putBytes $blockId")
```
again - what's the error here?

rxin commented on core/src/main/scala/org/apache/spark/storage/TachyonBlockManager.scala:
```diff
@@ -95,8 +94,29 @@ private[spark] class TachyonBlockManager() extends ExternalBlockManager with Log
+    try {
+      os.write(bytes.array())
+    } catch {
+      case e : Exception => 
```
can you change all the Exception to NonFatal(e)?

rxin commented on core/src/main/scala/org/apache/spark/storage/BlockManager.scala:
```diff
@@ -484,14 +485,15 @@ private[spark] class BlockManager(
-                }
+            val result = if (asBlockResult) {
+              externalBlockStore.getValues(blockId)
+              .map(new BlockResult(_, DataReadMethod.Memory, info.size))
```
indent here

rxin commented on core/src/main/scala/org/apache/spark/storage/BlockManager.scala:
```diff
@@ -1197,8 +1199,19 @@ private[spark] class BlockManager(
+   * Deserializes a InputStream into an iterator of values and disposes of it when the end of
+   * the iterator is reached.
+   */
+  def dataDeserializeStream(
```
dataDeserializeStream -> deserializeStream?

shimingfei replied:
use dataDeserializeStream to make it consistent with other code in the file.
there are dataSerialize and dataSerializeStream in the file.

rxin replied:
ok sounds good

Commit:
modify according to comments - format changes

AmplabJenkins commented:
Merged build triggered.

AmplabJenkins commented:
Merged build started.

SparkQA commented:
[Test build #32961 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/32961/consoleFull) for   PR 5908 at commit [`137f3d2`](https://github.com/apache/spark/commit/137f3d223c4c05269d00acf15b89cc0d4a0eb9e7).

shimingfei commented:
@rxin
 I have modified the code according to your comments, Thanks!

SparkQA commented:
[Test build #32961 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/32961/consoleFull) for   PR 5908 at commit [`137f3d2`](https://github.com/apache/spark/commit/137f3d223c4c05269d00acf15b89cc0d4a0eb9e7).
 * This patch **passes all tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/32961/
Test PASSed.

AmplabJenkins commented:
Merged build triggered.

AmplabJenkins commented:
Merged build started.

SparkQA commented:
[Test build #32977 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/32977/consoleFull) for   PR 5908 at commit [`0ace20e`](https://github.com/apache/spark/commit/0ace20ea3ce33b9f7c07491e32775643d0476519).

SparkQA commented:
[Test build #32977 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/32977/consoleFull) for   PR 5908 at commit [`0ace20e`](https://github.com/apache/spark/commit/0ace20ea3ce33b9f7c07491e32775643d0476519).
 * This patch **passes all tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/32977/
Test PASSed.

shimingfei commented:
@rxin 
Can this be merged ?

rxin commented on core/src/main/scala/org/apache/spark/storage/TachyonBlockManager.scala:
```diff
@@ -21,14 +21,13 @@ import java.io.IOException
 import org.apache.spark.{SparkException, SparkConf, Logging}
 import org.apache.spark.executor.ExecutorExitCode
 import org.apache.spark.util.Utils
+import scala.util.control.NonFatal
```
fix import ordering

rxin commented on core/src/main/scala/org/apache/spark/storage/TachyonBlockManager.scala:
```diff
@@ -95,8 +93,29 @@ private[spark] class TachyonBlockManager() extends ExternalBlockManager with Log
+      os.write(bytes.array())
+    } catch {
+      case NonFatal(e) => 
+        logWarning(s"Failed to put byes of block $blockId into Tachyon", e)
```
byTes

rxin commented on core/src/main/scala/org/apache/spark/storage/ExternalBlockStore.scala:
```diff
@@ -62,42 +62,72 @@ private[spark] class ExternalBlockStore(blockManager: BlockManager, executorId:
       } else {
-        logError(s"error in putBytes $blockId")
-        PutResult(bytes.limit(), null, Seq((blockId, BlockStatus.empty)))
+        logError(s"Error in putBytes $blockId : no ExternalBlockManager exists!")
```
```
s"Error in putBytes($blockId): no ExternalBlockManager has been configured."
```

rxin commented on core/src/main/scala/org/apache/spark/storage/ExternalBlockStore.scala:
```diff
@@ -62,42 +62,72 @@ private[spark] class ExternalBlockStore(blockManager: BlockManager, executorId:
+          blockId, Utils.bytesToString(size), finishTime - startTime))
+        PutResult(size, data)
+      } else {
+        logError(s"Error in putValues $blockId : no ExternalBlockManager exists!")
```
```
Error in putValues($blockId) : no ExternalBlockManager has been configured.
```

rxin commented on core/src/main/scala/org/apache/spark/storage/ExternalBlockStore.scala:
```diff
@@ -40,7 +40,7 @@ private[spark] class ExternalBlockStore(blockManager: BlockManager, executorId:
     } catch {
       case NonFatal(t) =>
-        logError(s"error in getSize from $blockId", t)
+        logError(s"Error in getSize from $blockId", t)
```
Can you change this to
```
Error calling getSize($blockId)
```

rxin replied:
and do this for the other messages too, since it is not clear "getSize", "getValues", ... are function calls. Let's just make them more explicitly.

rxin commented on core/src/main/scala/org/apache/spark/storage/BlockManager.scala:
```diff
@@ -485,14 +486,15 @@ private[spark] class BlockManager(
+              case Some(values) =>
+                return result
               case None =>
                 logDebug(s"Block $blockId not found in externalBlockStore")
```
```
Block $blockId not found in ExternalBlockStore
```

rxin commented on core/src/main/scala/org/apache/spark/storage/BlockManager.scala:
```diff
@@ -17,7 +17,8 @@
 package org.apache.spark.storage
 
-import java.io.{BufferedOutputStream, ByteArrayOutputStream, File, InputStream, OutputStream}
+import java.io.{BufferedOutputStream, BufferedInputStream, ByteArrayOutputStream, File,
```
use wildcard here since you already have 6 classes

rxin commented on core/src/main/scala/org/apache/spark/storage/ExternalBlockStore.scala:
```diff
@@ -62,42 +62,72 @@ private[spark] class ExternalBlockStore(blockManager: BlockManager, executorId:
+    try {
+      val startTime = System.currentTimeMillis
+      if (externalBlockManager.isDefined) {
+        // So that we do not modify the input offsets !
```
i think you can remove this 2 lines of comments. it is obvious what's happening

rxin commented on core/src/main/scala/org/apache/spark/storage/ExternalBlockStore.scala:
```diff
@@ -62,42 +62,72 @@ private[spark] class ExternalBlockStore(blockManager: BlockManager, executorId:
+        byteBuffer.rewind()
+        externalBlockManager.get.putBytes(blockId, byteBuffer)
+        val size = bytes.limit()
+        val data = if (returnValues) {
```
```
val data = if (returnValues) Right(bytes) else null
```

is block manager currently designed to return null, or is this new behavior?

shimingfei replied:
block manager currently designed to return null.

rxin commented on core/src/main/scala/org/apache/spark/storage/ExternalBlockStore.scala:
```diff
@@ -107,13 +137,19 @@ private[spark] class ExternalBlockStore(blockManager: BlockManager, executorId:
     } catch {
       case NonFatal(t) =>
-        logError(s"error in removing $blockId", t)
+        logError(s"Error in removing $blockId", t)
```
```Error in remove($blockId)```

rxin replied:
anyway please just change the logging messages to make them very explicit that they are function calls. We should either do this, or make them English sentences.

rxin commented on core/src/main/scala/org/apache/spark/storage/TachyonBlockManager.scala:
```diff
@@ -105,21 +124,35 @@ private[spark] class TachyonBlockManager() extends ExternalBlockManager with Log
     }
     val is = file.getInStream(ReadType.CACHE)
-    assert (is != null)
+    if (is == null) {
```
can you document when is would be null?

rxin commented on core/src/main/scala/org/apache/spark/storage/TachyonBlockManager.scala:
```diff
@@ -105,21 +124,35 @@ private[spark] class TachyonBlockManager() extends ExternalBlockManager with Log
+      return None
+    }
+    val is = file.getInStream(ReadType.CACHE)
+    if (is == null) {
```
this is 
```scala
Option(is).foreach { is =>
  blockManager.dataDeserializeStream(blockId, is)
}
```

Commit:
modify accroding to comments

shimingfei commented:
@rxin 
I have modified the code according to your comments, Thanks!

shimingfei commented on core/src/main/scala/org/apache/spark/storage/TachyonBlockManager.scala:
```diff
@@ -105,21 +128,31 @@ private[spark] class TachyonBlockManager() extends ExternalBlockManager with Log
       return None
     }
     val is = file.getInStream(ReadType.CACHE)
-    assert (is != null)
```
I checked Tachyon's code, it will not return null here

AmplabJenkins commented:
Merged build triggered.

AmplabJenkins commented:
Merged build started.

SparkQA commented:
[Test build #33073 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/33073/consoleFull) for   PR 5908 at commit [`033bc57`](https://github.com/apache/spark/commit/033bc57e83928d50e2d942027a51668ac39a580d).

SparkQA commented:
[Test build #33073 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/33073/consoleFull) for   PR 5908 at commit [`033bc57`](https://github.com/apache/spark/commit/033bc57e83928d50e2d942027a51668ac39a580d).
 * This patch **passes all tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/33073/
Test PASSed.

rxin commented:
cc @pwendell I will let you decide whether this should go into 1.4 or just master.

pwendell commented:
I think this will have to go into master branch since it is way past the deadline and people have already started testing this code. For more information about the deadline see - it is the same for every release:

http://apache-spark-developers-list.1001551.n3.nabble.com/Reminder-about-Spark-1-4-0-deadline-of-May-1st-td11832.html

pwendell commented:
Also this still needs an LGTM from @rxin or some active committer.

rxin commented:
LGTM.

shimingfei commented:
@pwendell 
can this be merged?
